### DISCUSSION TITLE
MQTT detection of BME 280 sensor

### DISCUSSION THREAD

**Discussion by TraxXx44** [fredlcore/BSB-LAN#700] [General]
Hey folks, Frederik,

Like many of us, I dived into home automation subject and started my first project of that type with the idea of making my heat pump smarter. After discovering dozens of "plug and play" devices which were actually not that smart, I came across Frederik's project and got in touch with him to order a BSB-LAN board. 

The first months following, I was a bit busy and didn't spend much time setting things up but I was rather thinking which platform I would use for my project, collecting informations on the pros and cons of each solution and also reading a lot of documentation to get familiar with the whole environnement because I'm not much into IoT systems and coding. 

I finally started the setup part few weeks ago and I'm already quite happy with the first results. AI helped a lot to understand how code works but it still has limits because the BSB-LAN project is very specific. After unsuccesful hours trying to solve my concern, I finally decided to post out there to ask for help or suggestions of alternative paths to explore.

To briefly sum up my configuration : 

**Hardware**

- Atlantic Alfea Excellia Duo 16 AI (heat pump)
- Raspberry Pi 4 Model B - 4 GB
- Olimex ESP32 EVB

**Software**

- Home Assistant OS
- Mosquitto Broker

My problem : 

I've connected a BME280 sensor to the right pins of the ESP 32 board extension. It's working fine as I can see temperature, humidity, pressure and altitude values directly in the BSB LAN interface. As it should be, the BME280 is correctly assigned to parameter 20200 with a value of 76, which correspond to its adress 0x76. 

![valeurs BME_BSB](https://github.com/user-attachments/assets/29d3e7ba-790c-4985-a6b6-ad07d4c0fcae)

However, as I believe, measurements are not assigned to the proper parameter numbers. From what I could see in the BSB_LAN_defs.h file, the measurements are supposed to be assigned to numbers 20201, 20202, 20203 and 20204 but they end up being assigned to 20200.1, 20200.2, 20200.3 and 20200.4 numbers. I noticed the code was written using floating numbers, adding up increments of 0.1 to the BME_280 value which is assigned to 20200.  It turns out only the 20200 entity is discovered by MQTT through Home Assistant automatic search. 

![BSB_lan_defs](https://github.com/user-attachments/assets/7bad6748-d9b5-4abd-a245-582c3a4c22b1)

--> I can easily find the different values of the BME280 using MQTT explorer

![mqtt_explorer](https://github.com/user-attachments/assets/fa0ca052-05ea-43f6-8c40-85b84158314c)

--> I've tried to reconfigure MQTT discovery from scratch but only the 20200 entity still shows up
--> I've tried to add them manually in the configuration.yaml file but the entity is not showing in the HA interface either
--> I've tried to replace 0.1 increments directly by integers in the BSB_LAN_defs.h file but as I could expect, that's not working and the code is broken doing this 

For now, my conclusion is that there is a problem for MQTT to recognize entities with decimal values since only 20200 is showing up but not the other values. 

Not sure this forum is the right place to discuss this type of issues but since it's mostly bsb-lan related, I thought that was relevant to do so. Also, I couldn't find any similar issues on my web searches and ChatGPT is out of ideas to solve this out. 

I hope you can help me on that matter :)

Thanks for reading !

**Comment by fredlcore** [comment]
Thanks for highlighting that, it's actually a bug that I'll have to fix. Will do asap.

**Comment by fredlcore** [comment]
I have removed the dot from the MQTT topic and that seems to fix the issue. Can you try it out on your side? If it doesn't work, please try removing retained topics using
`mosquitto_sub -h IP_ADDRESS_OF_MQTT_BROKER --remove-retained -t '#' -W 3`
several times.

**Reply by TraxXx44** [reply]
Hi Frederik,

Thanks for your assistance. Can you detail what you mean by "removing the dot from the MQTT topic" please ? At first, I thought about the dash from the topic prefix in the BSB-LAN parameters but I'm not sure you're actually talking about that. Where is the fix you're suggesting supposed to happen ?

**Reply by fredlcore** [reply]
You don't need to do anything other than install the most recent version from the master repository. In the new code, I have replaced any dot in the MQTT topic with a hyphen - and then it worked at least on my system, so I assume, HA doesn't like dots in the topic.

**Reply by TraxXx44** [reply]
Thank you for your help. I have downloaded the whole BSB-LAN folder again but first, had a look at the changes made since the version I flashed on my ESP32. 

I noticed many different files were updated and especially the _BSB_LAN_custom_defs.h.default_. From what I could read and see, changes on this file are minor and don't necessarily imply to update the file to have an operating ESP32. Updating that file would require that : 

- We go again in the process of sending you the device-specific parameter list to get the updated custom file 

OR 

- We compare the differences between the old custom version of this file (the first one we got from you) and the default new one and integrate ourselves the parts of the code that changed into the old file. 

This second option implies that changes are not so big in the structure and are enough noticable to just copy/paste the new parts of code into the former file. 

Am I right on this matter ?

**Reply by TraxXx44** [reply]
So I downloaded and flashed the last version of the repository on my ESP32, including the new BSB_LAN_config.h on which I updated the settings like in the older version. However, when I reached the BSB LAN web interface and clicked the sensors tab, BME280 is still recognized with 76 value but 20200.1, 20200.2 etc. remain as it was on the older version. What's worse here is that there are no more measurements displayed from temperature, humidity and pressure. It just mentions "query failed" next to a blank field...

**Comment by fredlcore** [comment]
First of all, the BSB_LAN_custom_defs.h file always stays the same, as already mentioned in the manual.
What I always find hard to understand is when people go to lenghts to try to analyze any differences in the code between their previous version, but without even mentioning any details about the version they have used before.
Of course, if you updated from a significantly earlier version (let's say 3.x), the settings will have to be done again, and that includes the BSB_LAN_config.h file, as is stated in the ChangeLog for breaking changes.

The thing is that 20200 is always returning a value once you set a pin number for the BME280 in BSB-LAN's settings, as this is just an enumeration, not an indication of a successful initialization. The "query failed" means that the sensor is not detected for whatever reason. If you have reasons to believe that this is a bug, then please go to the issues section and file a proper bug report, including all the relevant information that is listed there.

**Reply by TraxXx44** [reply]
For a reason I ignore, the value was set to 0x76 on my previous configuration _(#define MAJOR "4" - #define MINOR "2" - #define PATCH "44" - #define COMPILETIME "20241215161246")_ and it was working fine but not after updating with the new version of BSB-LAN. As a matter, I've changed all the values '0x76' in the ino file and replaced them with '0x77' and now the BME280 measurements are showing again in the BSB-LAN interface (no more "query failed" messages).

However, concerning MQTT, I haven't noticed any improvement concerning the values including dots. They still create issues and remain undetected by MQTT. Yet, I have flashed the new BSB-LAN code using both OTA and USB methods but strangely, dots are still showing like the code wasn't updated...

![valeurs BME_BSB](https://github.com/user-attachments/assets/77e8b1d2-6291-4e3d-a948-d67cb0a3a7c0)

Can you tell me in which file(s) you have replaced the dots with hyphens so I can investigate and see if I notice anything wrong ? 

Thank you,

**Reply by fredlcore** [reply]
I told you what to do, yet you ignore that, so I can't help you. Also, I didn't change the dot in the actual values, of course, I changed them in the MQTT topic, as mentioned above.

**Reply by TraxXx44** [reply]
I'll submit a proper bug report as soon as I can. And sorry if my questions might sound silly but despite all the documentation I've read until now, IoT and all related vocabulary remains a whole new subject to me so I'm not very familiar with it. For instance, when you mention "In the new code, I have replaced any dot in the MQTT topic with a hyphen", I have no clue where you did that. Could you just tell me so I can understand better just on this specific action ? 
Thank you

**Reply by fredlcore** [reply]
You can just look at the GitHub commit history where it will show you exactly what changes were made. The reason why I'm not entertaining such requests is that I want to fix any bug. If you cannot contribute anything to the bugfixing, then all you can - and need - to do is to provide me with the information that I need. If you rather want to spend your time learning where and what I'm doing, that's fine with me, but then I will attend to other issues.

**Comment by fredlcore** [comment]
Another user has filed a bug report here https://github.com/fredlcore/BSB-LAN/issues/703 which showed the same kind of problems and he confirmed that my fix solved the problem, so I unless you provide additional context, I assume that this was the same problem here.

**Reply by TraxXx44** [reply]
I've made a clean re-install of Home Assistant and everything is working fine now. BME280 values - among other ones - are now showing up in MQTT entities. Thanks for fixing Frederik and next time I'm coming across what looks like a bug, I'll fill a proper bug report !
